### PR TITLE
CP layout integration

### PR DESCRIPTION
**Pull request by Mellonta**
## Background

TE softmax attention requires a zigzag CP layout, while linear attention does not. With
chunkwise CP, using the zigzag layout for linear-attention layers doubles their communication. This PR wires the `megatron/core/context_parallel` module from #6764 into hybrid models.

## Summary
- Add `linear_cp_layout` and `attention_cp_layout` to `TransformerConfig`, both defaulting to `zigzag`, so existing runs are unchanged.
- Shard the batch into contiguous CP slices when the linear layout is contiguous, keeping the dense attention mask in zigzag.

## Tests
```
uv run python -m torch.distributed.run --nproc-per-node 4 -m pytest -q \
    tests/unit_tests/data/test_get_batch.py
```

### CODE REVIEW

File: megatron/core/transformer/transformer_config.py [hunk 1]
```diff
@@ -1046,6 +1047,12 @@ class TransformerConfig(ModelParallelConfig):
     and P2P communications in high-level CP groups (e.g., via IBLink).
     """
 
+    linear_cp_layout: CPLayout = "zigzag"
+    """CP layout for linear-attention layers."""
+
+    attention_cp_layout: CPLayout = "zigzag"
+    """CP layout for softmax-attention layers."""
+
```
Comment (Phlip79) [inline megatron/core/transformer/transformer_config.py:1055]: Will this be problematic setting the default to zigzag?

File: megatron/core/utils.py [hunk 2]
```diff
@@ -2562,6 +2562,49 @@ def _get_batch_on_this_cp_rank_per_sequence_balancing(
     return batch
 
 
+def _get_batch_on_this_cp_rank_contiguous(
```
Comment (Phlip79) [inline megatron/core/utils.py]: This file is already way too big (not your fault).

Can you move this function to `megatron/core/context_parallel/utils.py` or similar?

File: megatron/core/transformer/transformer_config.py (hunk 1 above)
Comment (Mellonta) [reply megatron/core/transformer/transformer_config.py:1055]: No, it's for backward compatibility. Zigzag is what Megatron does today.

**Review by Phlip79 — APPROVED** [review]

File: megatron/core/transformer/transformer_config.py (hunk 1 above)
Comment (Phlip79) [reply megatron/core/transformer/transformer_config.py:1055]: Got it, thanks.

**Comment by Phlip79** [comment]
Please update PR description as well

**Review by deepakn94 — APPROVED** [review]

**Review by zhongbozhu — APPROVED** [review]
LGTM
